File: ingestion/shared/jurisdictions.ts
/**
 * 団体の同一性。**層に依存しない。**
 *
 * ①予算・②調達・③会議録はすべて全国地方公共団体コードで束ねるので、
 * 名称と識別子はどの層からも参照される。
 *
 * ⚠️ **以前はこれが③会議録のゲート判定ファイルに同居していた。**
 * そのため予算の粒度調査が、母集団を得るためだけに③のファイルを読んでいた。
 * 設計上は分離していたのに、物理構造では③が落ちると①も動かない状態だった。
 */
import { z } from 'zod'

export const Jurisdiction = z.object({
  name: z.string(),
  /** Open Civic Data 形式の識別子。本家未登録の市区町村は同形式で自前定義 */
  ocdId: z.string().regex(/^ocd-division\/country:jp\/prefecture:\d{2}\/city:\d{6}$/),
  /** 東京都カタログでの公開データセット数。母集団の目安であって、粒度の判定には使わない */
  tokyoCatalogDatasets: z.number().int().nonnegative().nullable(),
})

export const Registry = z.object({
  note: z.string(),
  generatedAt: z.iso.date().optional(),
  /** キーは全国地方公共団体コード（6桁） */
  jurisdictions: z.record(z.string().regex(/^\d{6}$/), Jurisdiction),
})

export type Jurisdiction = z.infer<typeof Jurisdiction>
export type Registry = z.infer<typeof Registry>

export const REGISTRY_PATH = new URL('../../data/shared/jurisdictions.json', import.meta.url).pathname

export async function loadJurisdictions(): Promise<Registry> {
  return Registry.parse(JSON.parse(await Bun.file(REGISTRY_PATH).text()))
}

File: ingestion/shared/jurisdictions.py
"""団体の同一性。**層に依存しない。** `jurisdictions.ts` の Python 版対。

①予算・②調達・③会議録はすべて全国地方公共団体コードで束ねるので、名称と識別子は
どの層からも参照される。②③は TS からこの JSON を読んでいるが、①予算は Python
（`ingestion/budget/*`）なので、同じ事実を Python 側で再宣言せずここから引く。

⚠️ 以前は `ingestion/budget/sources.toml` が `jurisdiction_name` を団体×年度ごとに
反復宣言しており（狛江市だけで6回）、`jurisdictions.json` と突き合わせる経路が無かった。
sources.toml に誤記があっても検知されず、配布物の `jurisdiction_label` へそのまま出ていた。
"""

from __future__ import annotations

import json
from pathlib import Path

REGISTRY_PATH = Path(__file__).resolve().parent / "jurisdictions.json"


def load_jurisdictions(path: Path = REGISTRY_PATH) -> dict[str, dict]:
    """全国地方公共団体コードをキーにした団体情報（name / ocdId / tokyoCatalogDatasets）"""
    return json.loads(path.read_text(encoding="utf-8"))["jurisdictions"]


def jurisdiction_name(code: str, path: Path = REGISTRY_PATH) -> str:
    """団体コードから名称を引く。**登録が無ければ例外**（黙って欠けさせない）"""
    registry = load_jurisdictions(path)
    if code not in registry:
        raise KeyError(
            f"団体コード「{code}」が ingestion/shared/jurisdictions.json に無い。"
            f"団体の名称と識別子はそこで一元管理している（AGENTS.md 参照）"
        )
    return registry[code]["name"]
